Upgrade to Hibernate ORM 7.0.0.Beta5

It would seem the latest version of Hibernate Reactive 3 is not compatible. Getting things like this in my Quarkus upgrade PR (https://github.com/quarkusio/quarkus/pull/41310):

```
2025-03-24T09:35:16.1351507Z 2025-03-24 09:35:16,124 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-2) HTTP Request to /tests/reactiveRemoveTransientEntity failed, error id: 701e6fd2-1ea4-4083-a920-ec5eb60c3125-2: org.hibernate.HibernateException: java.util.concurrent.CompletionException: java.lang.NoSuchMethodError: 'void org.hibernate.engine.internal.Nullability.checkNullability(java.lang.Object[], org.hibernate.persister.entity.EntityPersister, org.hibernate.engine.internal.Nullability$NullabilityCheckType)'
2025-03-24T09:35:16.1355894Z 	at org.hibernate.reactive.session.impl.ReactiveExceptionConverter.convert(ReactiveExceptionConverter.java:28)
2025-03-24T09:35:16.1357855Z 	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.lambda$fireRemove$17(ReactiveSessionImpl.java:850)
2025-03-24T09:35:16.1359756Z 	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
2025-03-24T09:35:16.1361181Z 	at java.base/java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:950)
2025-03-24T09:35:16.1362715Z 	at java.base/java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2340)
2025-03-24T09:35:16.1364000Z 	at java.base/java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:144)
2025-03-24T09:35:16.1365549Z 	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fireRemove(ReactiveSessionImpl.java:839)
2025-03-24T09:35:16.1367134Z 	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveRemove(ReactiveSessionImpl.java:789)
2025-03-24T09:35:16.1368707Z 	at org.hibernate.reactive.mutiny.impl.MutinySessionImpl.lambda$remove$12(MutinySessionImpl.java:267)
2025-03-24T09:35:16.1370696Z 	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
2025-03-24T09:35:16.1372844Z 	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
2025-03-24T09:35:16.1374909Z 	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
2025-03-24T09:35:16.1392018Z 	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
2025-03-24T09:35:16.1393221Z 	at org.hibernate.reactive.context.impl.VertxContext.execute(VertxContext.java:91)
2025-03-24T09:35:16.1394339Z 	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.subscribe(UniRunSubscribeOn.java:25)
2025-03-24T09:35:16.1395404Z 	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
2025-03-24T09:35:16.1396977Z 	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
2025-03-24T09:35:16.1399332Z 	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
2025-03-24T09:35:16.1401129Z 	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:47)
2025-03-24T09:35:16.1402856Z 	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forwardResult(UniCreateFromCompletionStage.java:63)
2025-03-24T09:35:16.1410392Z 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
2025-03-24T09:35:16.1411691Z 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
2025-03-24T09:35:16.1412816Z 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
2025-03-24T09:35:16.1413938Z 	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
2025-03-24T09:35:16.1415182Z 	at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:131)
2025-03-24T09:35:16.1416634Z 	at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.lambda$unroll$0(AsyncTrampoline.java:126)
2025-03-24T09:35:16.1418016Z 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
2025-03-24T09:35:16.1419268Z 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
2025-03-24T09:35:16.1420855Z 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
2025-03-24T09:35:16.1422015Z 	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
2025-03-24T09:35:16.1422945Z 	at io.vertx.core.Future.lambda$toCompletionStage$3(Future.java:602)
2025-03-24T09:35:16.1423715Z 	at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
2025-03-24T09:35:16.1424557Z 	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
2025-03-24T09:35:16.1425413Z 	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
2025-03-24T09:35:16.1426373Z 	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:88)
2025-03-24T09:35:16.1427449Z 	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:32)
2025-03-24T09:35:16.1428311Z 	at io.vertx.core.Promise.complete(Promise.java:66)
2025-03-24T09:35:16.1429029Z 	at io.vertx.core.Promise.handle(Promise.java:51)
2025-03-24T09:35:16.1429752Z 	at io.vertx.core.Promise.handle(Promise.java:29)
2025-03-24T09:35:16.1430476Z 	at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
2025-03-24T09:35:16.1431313Z 	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
2025-03-24T09:35:16.1432190Z 	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
2025-03-24T09:35:16.1433226Z 	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
2025-03-24T09:35:16.1434296Z 	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
2025-03-24T09:35:16.1435434Z 	at io.vertx.sqlclient.impl.TransactionImpl.lambda$wrap$0(TransactionImpl.java:72)
2025-03-24T09:35:16.1436534Z 	at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
2025-03-24T09:35:16.1437654Z 	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:60)
2025-03-24T09:35:16.1438934Z 	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:312)
2025-03-24T09:35:16.1440188Z 	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:168)
2025-03-24T09:35:16.1441284Z 	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:57)
2025-03-24T09:35:16.1442344Z 	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
2025-03-24T09:35:16.1443190Z 	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
2025-03-24T09:35:16.1444012Z 	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
2025-03-24T09:35:16.1444929Z 	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
2025-03-24T09:35:16.1445905Z 	at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
2025-03-24T09:35:16.1446993Z 	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:324)
2025-03-24T09:35:16.1448150Z 	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:137)
2025-03-24T09:35:16.1449065Z 	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:342)
2025-03-24T09:35:16.1449881Z 	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
2025-03-24T09:35:16.1450659Z 	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:389)
2025-03-24T09:35:16.1451591Z 	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
2025-03-24T09:35:16.1452429Z 	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
2025-03-24T09:35:16.1453574Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
2025-03-24T09:35:16.1455046Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
2025-03-24T09:35:16.1456481Z 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
2025-03-24T09:35:16.1458108Z 	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
2025-03-24T09:35:16.1459504Z 	at io.vertx.db2client.impl.codec.DB2Encoder.lambda$write$0(DB2Encoder.java:72)
2025-03-24T09:35:16.1460949Z 	at io.vertx.db2client.impl.codec.ExtendedQueryCommandCodec.decodeQuery(ExtendedQueryCommandCodec.java:52)
2025-03-24T09:35:16.1462322Z 	at io.vertx.db2client.impl.codec.QueryCommandBaseCodec.decodePayload(QueryCommandBaseCodec.java:69)
2025-03-24T09:35:16.1463425Z 	at io.vertx.db2client.impl.codec.DB2Decoder.decodePayload(DB2Decoder.java:79)
2025-03-24T09:35:16.1464322Z 	at io.vertx.db2client.impl.codec.DB2Decoder.decode(DB2Decoder.java:52)
2025-03-24T09:35:16.1465480Z 	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530)
2025-03-24T09:35:16.1466789Z 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469)
2025-03-24T09:35:16.1467893Z 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
2025-03-24T09:35:16.1469260Z 	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
2025-03-24T09:35:16.1470837Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
2025-03-24T09:35:16.1472291Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
2025-03-24T09:35:16.1473715Z 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
2025-03-24T09:35:16.1475048Z 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
2025-03-24T09:35:16.1476413Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
2025-03-24T09:35:16.1477863Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
2025-03-24T09:35:16.1479185Z 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
2025-03-24T09:35:16.1480569Z 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
2025-03-24T09:35:16.1481678Z 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
2025-03-24T09:35:16.1482735Z 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
2025-03-24T09:35:16.1483829Z 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
2025-03-24T09:35:16.1484846Z 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
2025-03-24T09:35:16.1485904Z 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
2025-03-24T09:35:16.1486944Z 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
2025-03-24T09:35:16.1487940Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2025-03-24T09:35:16.1488780Z 	at java.base/java.lang.Thread.run(Thread.java:840)
2025-03-24T09:35:16.1491027Z Caused by: java.util.concurrent.CompletionException: java.lang.NoSuchMethodError: 'void org.hibernate.engine.internal.Nullability.checkNullability(java.lang.Object[], org.hibernate.persister.entity.EntityPersister, org.hibernate.engine.internal.Nullability$NullabilityCheckType)'
2025-03-24T09:35:16.1500511Z 	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
2025-03-24T09:35:16.1502027Z 	at java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:761)
2025-03-24T09:35:16.1503549Z 	at java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:735)
2025-03-24T09:35:16.1504959Z 	at java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2182)
2025-03-24T09:35:16.1506340Z 	at java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:144)
2025-03-24T09:35:16.1508784Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.lambda$deleteEntity$5(DefaultReactiveDeleteEventListener.java:438)
2025-03-24T09:35:16.1511223Z 	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
2025-03-24T09:35:16.1512780Z 	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
2025-03-24T09:35:16.1514178Z 	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
2025-03-24T09:35:16.1515950Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.deleteEntity(DefaultReactiveDeleteEventListener.java:431)
2025-03-24T09:35:16.1518132Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.delete(DefaultReactiveDeleteEventListener.java:282)
2025-03-24T09:35:16.1520303Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.deletePersistentInstance(DefaultReactiveDeleteEventListener.java:268)
2025-03-24T09:35:16.1521672Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.delete(DefaultReactiveDeleteEventListener.java:133)
2025-03-24T09:35:16.1522849Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.lambda$fetchAndDelete$0(DefaultReactiveDeleteEventListener.java:203)
2025-03-24T09:35:16.1523734Z 	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
2025-03-24T09:35:16.1524396Z 	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
2025-03-24T09:35:16.1525016Z 	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
2025-03-24T09:35:16.1525833Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.fetchAndDelete(DefaultReactiveDeleteEventListener.java:203)
2025-03-24T09:35:16.1526854Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.reactiveOnDelete(DefaultReactiveDeleteEventListener.java:124)
2025-03-24T09:35:16.1527875Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.reactiveOnDelete(DefaultReactiveDeleteEventListener.java:111)
2025-03-24T09:35:16.1528837Z 	at org.hibernate.event.service.internal.EventListenerGroupImpl.lambda$fireEventOnEachListener$0(EventListenerGroupImpl.java:166)
2025-03-24T09:35:16.1529918Z 	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
2025-03-24T09:35:16.1531199Z 	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
2025-03-24T09:35:16.1532382Z 	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
2025-03-24T09:35:16.1534218Z 	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:166)
2025-03-24T09:35:16.1535839Z 	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fireRemove(ReactiveSessionImpl.java:838)
2025-03-24T09:35:16.1537684Z 	... 84 more
2025-03-24T09:35:16.1539382Z Caused by: java.lang.NoSuchMethodError: 'void org.hibernate.engine.internal.Nullability.checkNullability(java.lang.Object[], org.hibernate.persister.entity.EntityPersister, org.hibernate.engine.internal.Nullability$NullabilityCheckType)'
2025-03-24T09:35:16.1542510Z 	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.lambda$deleteEntity$4(DefaultReactiveDeleteEventListener.java:439)
2025-03-24T09:35:16.1545162Z 	at java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:757)
2025-03-24T09:35:16.1546007Z 	... 108 more
```

--- a/gradle.properties
+++ b/gradle.properties
@@ -35,7 +35,7 @@ org.gradle.java.installations.auto-download=false
 #enableMavenLocalRepo = true
 
 # The default Hibernate ORM version (override using `-PhibernateOrmVersion=the.version.you.want`)
-hibernateOrmVersion = 7.0.0.Beta4
+hibernateOrmVersion = 7.0.0.Beta5
 
 # Override default Hibernate ORM Gradle plugin version
 # Using the stable version because I don't know how to configure the build to download the snapshot version from

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveDeleteEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveDeleteEventListener.java
@@ -436,11 +436,8 @@ public class DefaultReactiveDeleteEventListener
 						persister
 				).nullifyTransientReferences( entityEntry.getDeletedState() )
 						.thenAccept( vv -> {
-							new Nullability( session ).checkNullability(
-									entityEntry.getDeletedState(),
-									persister,
-									Nullability.NullabilityCheckType.DELETE
-							);
+							new Nullability( session, Nullability.NullabilityCheckType.DELETE )
+									.checkNullability( entityEntry.getDeletedState(), persister );
 							persistenceContext.registerNullifiableEntityKey( key );
 
 							final ReactiveActionQueue actionQueue = actionQueue( session );

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveEntityBatchLoaderArrayParam.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveEntityBatchLoaderArrayParam.java
@@ -10,7 +10,6 @@ import java.util.Locale;
 import java.util.concurrent.CompletionStage;
 
 import org.hibernate.LockOptions;
-import org.hibernate.engine.internal.BatchFetchQueueHelper;
 import org.hibernate.engine.spi.EntityKey;
 import org.hibernate.engine.spi.LoadQueryInfluencers;
 import org.hibernate.engine.spi.SessionFactoryImplementor;
@@ -160,7 +159,8 @@ public class ReactiveEntityBatchLoaderArrayParam<T> extends ReactiveSingleIdEnti
 							continue;
 						}
 						// found or not, remove the key from the batch-fetch queue
-						BatchFetchQueueHelper.removeBatchLoadableEntityKey( id, getLoadable(), session );
+						session.getPersistenceContextInternal().getBatchFetchQueue()
+								.removeBatchLoadableEntityKey( session.generateEntityKey( id, getLoadable().getEntityPersister() ) );
 					}
 				} );
 	}

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveMultiIdEntityLoaderArrayParam.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveMultiIdEntityLoaderArrayParam.java
@@ -13,7 +13,6 @@ import java.util.concurrent.CompletionStage;
 
 import org.hibernate.LockMode;
 import org.hibernate.LockOptions;
-import org.hibernate.engine.internal.BatchFetchQueueHelper;
 import org.hibernate.engine.spi.BatchFetchQueue;
 import org.hibernate.engine.spi.EntityEntry;
 import org.hibernate.engine.spi.EntityKey;
@@ -206,7 +205,7 @@ public class ReactiveMultiIdEntityLoaderArrayParam<E> extends ReactiveAbstractMu
 						// the element value at this position in the result List should be
 						// the EntityKey for that entity - reuse it
 						final EntityKey entityKey = (EntityKey) result.get( resultIndex );
-						BatchFetchQueueHelper.removeBatchLoadableEntityKey( entityKey, session );
+						session.getPersistenceContextInternal().getBatchFetchQueue().removeBatchLoadableEntityKey( entityKey );
 						Object entity = persistenceContext.getEntity( entityKey );
 						if ( entity != null && !loadOptions.isReturnOfDeletedEntitiesEnabled() ) {
 							// make sure it is not DELETED
@@ -293,7 +292,8 @@ public class ReactiveMultiIdEntityLoaderArrayParam<E> extends ReactiveAbstractMu
 					continue;
 				}
 				// found or not, remove the key from the batch-fetch queue
-				BatchFetchQueueHelper.removeBatchLoadableEntityKey( id, getLoadable(), session );
+				session.getPersistenceContextInternal().getBatchFetchQueue()
+						.removeBatchLoadableEntityKey( session.generateEntityKey( id, getLoadable().getEntityPersister() ) );
 			}
 
 			return result;

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinySessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinySessionImpl.java
@@ -173,7 +173,7 @@ public class MutinySessionImpl implements Mutiny.Session {
 
 	@Override
 	public <R> Query<R> createNamedQuery(String queryName) {
-		return new MutinyQueryImpl<>( delegate.createReactiveNamedQuery( queryName, null ), factory );
+		return new MutinyQueryImpl<>( delegate.createReactiveNamedQuery( queryName ), factory );
 	}
 
 	@Override

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinyStatelessSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinyStatelessSessionImpl.java
@@ -93,7 +93,7 @@ public class MutinyStatelessSessionImpl implements Mutiny.StatelessSession {
 
 	@Override
 	public <R> Query<R> createNamedQuery(String queryName) {
-		return new MutinyQueryImpl<>( delegate.createReactiveNamedQuery( queryName, null ), factory );
+		return new MutinyQueryImpl<>( delegate.createReactiveNamedQuery( queryName ), factory );
 	}
 
 	@Override

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/query/ReactiveQueryImplementor.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/query/ReactiveQueryImplementor.java
@@ -5,7 +5,6 @@
  */
 package org.hibernate.reactive.query;
 
-import java.io.Serializable;
 import java.time.Instant;
 import java.util.Calendar;
 import java.util.Collection;
@@ -22,12 +21,6 @@ import jakarta.persistence.Parameter;
 import jakarta.persistence.TemporalType;
 
 public interface ReactiveQueryImplementor<R> extends ReactiveQuery<R> {
-
-	void setOptionalId(Serializable id);
-
-	void setOptionalEntityName(String entityName);
-
-	void setOptionalObject(Object optionalObject);
 
 	QueryParameterBindings getParameterBindings();
 

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/query/sql/internal/ReactiveNativeQueryImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/query/sql/internal/ReactiveNativeQueryImpl.java
@@ -26,7 +26,6 @@ import org.hibernate.engine.spi.SharedSessionContractImplementor;
 import org.hibernate.graph.GraphSemantic;
 import org.hibernate.graph.RootGraph;
 import org.hibernate.graph.spi.RootGraphImplementor;
-import org.hibernate.internal.AbstractSharedSessionContract;
 import org.hibernate.metamodel.model.domain.BasicDomainType;
 import org.hibernate.query.BindableType;
 import org.hibernate.query.Order;
@@ -65,8 +64,18 @@ public class ReactiveNativeQueryImpl<R> extends NativeQueryImpl<R>
 
 	private final ReactiveAbstractSelectionQuery<R> selectionQueryDelegate;
 
-	public ReactiveNativeQueryImpl(String memento, SharedSessionContractImplementor session) {
-		super( memento, session );
+	public ReactiveNativeQueryImpl(String sql, SharedSessionContractImplementor session) {
+		super( sql, null, session );
+		this.selectionQueryDelegate = createSelectionQueryDelegate( session );
+	}
+
+	public ReactiveNativeQueryImpl(String sql, Class<R> resultClass, SharedSessionContractImplementor session) {
+		super( sql, resultClass, session );
+		this.selectionQueryDelegate = createSelectionQueryDelegate( session );
+	}
+
+	public ReactiveNativeQueryImpl(String sql, NamedResultSetMappingMemento resultSetMappingMemento, Class<R> resultClass, SharedSessionContractImplementor session) {
+		super( sql, resultSetMappingMemento, resultClass, session);
 		this.selectionQueryDelegate = createSelectionQueryDelegate( session );
 	}
 
@@ -88,14 +97,6 @@ public class ReactiveNativeQueryImpl<R> extends NativeQueryImpl<R>
 			String resultSetMappingName,
 			SharedSessionContractImplementor session) {
 		super( memento, resultSetMappingName, session );
-		this.selectionQueryDelegate = createSelectionQueryDelegate( session );
-	}
-
-	public ReactiveNativeQueryImpl(
-			String sqlString,
-			NamedResultSetMappingMemento resultSetMappingMemento,
-			AbstractSharedSessionContract session) {
-		super( sqlString, resultSetMappingMemento, session );
 		this.selectionQueryDelegate = createSelectionQueryDelegate( session );
 	}
 

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/query/sqm/internal/ReactiveSimpleDeleteQueryPlan.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/query/sqm/internal/ReactiveSimpleDeleteQueryPlan.java
@@ -37,9 +37,7 @@ import org.hibernate.spi.NavigablePath;
 import org.hibernate.sql.ast.SqlAstTranslator;
 import org.hibernate.sql.ast.tree.AbstractUpdateOrDeleteStatement;
 import org.hibernate.sql.ast.tree.MutationStatement;
-import org.hibernate.sql.ast.tree.expression.ColumnReference;
 import org.hibernate.sql.ast.tree.expression.Expression;
-import org.hibernate.sql.ast.tree.expression.JdbcLiteral;
 import org.hibernate.sql.ast.tree.from.MutatingTableReferenceGroupWrapper;
 import org.hibernate.sql.ast.tree.from.NamedTableReference;
 import org.hibernate.sql.ast.tree.predicate.InSubQueryPredicate;
@@ -95,29 +93,28 @@ public class ReactiveSimpleDeleteQueryPlan extends SimpleDeleteQueryPlan impleme
 		return factory.getJdbcServices()
 				.getJdbcEnvironment()
 				.getSqlAstTranslatorFactory()
-				.buildMutationTranslator( factory, mutationStatement() );
+				.buildMutationTranslator( factory, createDeleteAst() );
 	}
 
-	private MutationStatement mutationStatement() {
+	// Copy and paste from superclass
+	private MutationStatement createDeleteAst() {
+		final MutationStatement ast;
 		if ( entityDescriptor.getSoftDeleteMapping() == null ) {
-			return sqmInterpretation.getSqlAst();
+			ast = sqmInterpretation.getSqlAst();
 		}
-		final AbstractUpdateOrDeleteStatement sqlDeleteAst = sqmInterpretation.getSqlAst();
-		final NamedTableReference targetTable = sqlDeleteAst.getTargetTable();
-		final SoftDeleteMapping columnMapping = getEntityDescriptor().getSoftDeleteMapping();
-		final ColumnReference columnReference = new ColumnReference( targetTable, columnMapping );
-		//noinspection rawtypes,unchecked
-		final JdbcLiteral jdbcLiteral = new JdbcLiteral(
-				columnMapping.getDeletedLiteralValue(),
-				columnMapping.getJdbcMapping()
-		);
-		final Assignment assignment = new Assignment( columnReference, jdbcLiteral );
+		else {
+			final AbstractUpdateOrDeleteStatement sqlDeleteAst = sqmInterpretation.getSqlAst();
+			final NamedTableReference targetTable = sqlDeleteAst.getTargetTable();
+			final SoftDeleteMapping columnMapping = getEntityDescriptor().getSoftDeleteMapping();
+			final Assignment assignment = columnMapping.createSoftDeleteAssignment( targetTable );
 
-		return new UpdateStatement(
-				targetTable,
-				Collections.singletonList( assignment ),
-				sqlDeleteAst.getRestriction()
-		);
+			ast = new UpdateStatement(
+					targetTable,
+					Collections.singletonList( assignment ),
+					sqlDeleteAst.getRestriction()
+			);
+		}
+		return ast;
 	}
 
 	@Override

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/ReactiveQueryProducer.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/ReactiveQueryProducer.java
@@ -57,6 +57,8 @@ public interface ReactiveQueryProducer extends ReactiveConnectionSupplier {
 
 	<R> ReactiveQuery<R> createReactiveQuery(String queryString, Class<R> resultType);
 
+	<R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String queryString);
+
 	<R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String queryString, Class<R> resultType);
 
 	<R> ReactiveNativeQuery<R> createReactiveNativeQuery(String sqlString);

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java
@@ -72,13 +72,16 @@ import org.hibernate.pretty.MessageHelper;
 import org.hibernate.proxy.HibernateProxy;
 import org.hibernate.proxy.LazyInitializer;
 import org.hibernate.query.IllegalMutationQueryException;
+import org.hibernate.query.UnknownNamedQueryException;
 import org.hibernate.query.criteria.JpaCriteriaInsert;
 import org.hibernate.query.hql.spi.SqmQueryImplementor;
 import org.hibernate.query.named.NamedResultSetMappingMemento;
 import org.hibernate.query.spi.HqlInterpretation;
 import org.hibernate.query.spi.QueryImplementor;
+import org.hibernate.query.sql.spi.NamedNativeQueryMemento;
 import org.hibernate.query.sql.spi.NativeQueryImplementor;
 import org.hibernate.query.sqm.internal.SqmUtil;
+import org.hibernate.query.sqm.spi.NamedSqmQueryMemento;
 import org.hibernate.query.sqm.tree.SqmStatement;
 import org.hibernate.query.sqm.tree.delete.SqmDeleteStatement;
 import org.hibernate.query.sqm.tree.insert.SqmInsertStatement;
@@ -131,7 +134,6 @@ import static org.hibernate.engine.internal.ManagedTypeHelper.isPersistentAttrib
 import static org.hibernate.engine.spi.NaturalIdResolutions.INVALID_NATURAL_ID_REFERENCE;
 import static org.hibernate.event.spi.LoadEventListener.IMMEDIATE_LOAD;
 import static org.hibernate.internal.util.StringHelper.isEmpty;
-import static org.hibernate.internal.util.StringHelper.isNotEmpty;
 import static org.hibernate.proxy.HibernateProxy.extractLazyInitializer;
 import static org.hibernate.reactive.common.InternalStateAssertions.assertUseOnEventLoop;
 import static org.hibernate.reactive.persister.entity.impl.ReactiveEntityPersister.forceInitialize;
@@ -443,14 +445,16 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 
 	@Override @Deprecated(forRemoval = true)
 	public <R> ReactiveNativeQuery<R> createReactiveNativeQuery(String sqlString, String resultSetMappingName) {
+		if ( isEmpty( resultSetMappingName ) ) {
+			throw new IllegalArgumentException( "Result set mapping name was not specified" );
+		}
+
 		checkOpen();
 		pulseTransactionCoordinator();
 		delayedAfterCompletion();
 
 		try {
-			return isNotEmpty( resultSetMappingName )
-					? new ReactiveNativeQueryImpl<>( sqlString, getResultSetMappingMemento( resultSetMappingName ), this )
-					: new ReactiveNativeQueryImpl<>( sqlString, this );
+			return new ReactiveNativeQueryImpl<>( sqlString, getResultSetMappingMemento( resultSetMappingName ), null, this );
 			//TODO: why no applyQuerySettingsAndHints( query ); ???
 		}
 		catch (RuntimeException he) {
@@ -498,8 +502,77 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 	}
 
 	@Override
+	public <R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String name) {
+		checksBeforeQueryCreation();
+		try {
+			return (ReactiveQueryImplementor<R>) buildNamedQuery(
+					name,
+					this::createSqmQueryImplementor,
+					this::createNativeQueryImplementor
+			);
+		}
+		catch (RuntimeException e) {
+			throw convertNamedQueryException( e );
+		}
+	}
+
+	@Override
 	public <R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String name, Class<R> resultType) {
-		return (ReactiveQueryImplementor<R>) buildNamedQuery( name, resultType );
+		checksBeforeQueryCreation();
+		if ( resultType == null ) {
+			throw new IllegalArgumentException( "Result class is null" );
+		}
+		try {
+			return buildNamedQuery(
+					name,
+					memento -> createReactiveSqmQueryImplementor( resultType, memento ),
+					memento -> createReactiveNativeQueryImplementor( resultType, memento )
+			);
+		}
+		catch (RuntimeException e) {
+			throw convertNamedQueryException( e );
+		}
+	}
+
+	private void checksBeforeQueryCreation() {
+		checkOpen();
+		checkTransactionSynchStatus();
+	}
+
+	protected <T> ReactiveNativeQueryImpl<T> createReactiveNativeQueryImplementor(Class<T> resultType, NamedNativeQueryMemento<?> memento) {
+		final NativeQueryImplementor<T> query = memento.toQuery(this, resultType );
+		if ( isEmpty( query.getComment() ) ) {
+			query.setComment( "dynamic native SQL query" );
+		}
+		applyQuerySettingsAndHints( query );
+		return (ReactiveNativeQueryImpl<T>) query;
+	}
+
+	protected <T> ReactiveQuerySqmImpl<T> createReactiveSqmQueryImplementor(Class<T> resultType, NamedSqmQueryMemento<?> memento) {
+		final SqmQueryImplementor<T> query = memento.toQuery( this, resultType );
+		if ( isEmpty( query.getComment() ) ) {
+			query.setComment( "dynamic query" );
+		}
+		applyQuerySettingsAndHints( query );
+		if ( memento.getLockOptions() != null ) {
+			query.setLockOptions( memento.getLockOptions() );
+		}
+		return (ReactiveQuerySqmImpl<T>) query;
+	}
+
+	private RuntimeException convertNamedQueryException(RuntimeException e) {
+		if ( e instanceof UnknownNamedQueryException ) {
+			// JPA expects this to mark the transaction for rollback only
+			getTransactionCoordinator().getTransactionDriverControl().markRollbackOnly();
+			// it also expects an IllegalArgumentException, so wrap UnknownNamedQueryException
+			return new IllegalArgumentException( e.getMessage(), e );
+		}
+		else if ( e instanceof IllegalArgumentException ) {
+			return e;
+		}
+		else {
+			return getExceptionConverter().convert( e );
+		}
 	}
 
 	@Override
@@ -584,7 +657,7 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 		delayedAfterCompletion();
 
 		try {
-			final ReactiveNativeQueryImpl<R> query = new ReactiveNativeQueryImpl<>( queryString, this );
+			final ReactiveNativeQueryImpl<R> query = new ReactiveNativeQueryImpl<>( queryString, null, this );
 			addAffectedEntities( affectedEntities, query );
 			if ( isEmpty( query.getComment() ) ) {
 				query.setComment( "dynamic native SQL query" );
@@ -622,12 +695,11 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 		checkOpen();
 		pulseTransactionCoordinator();
 		delayedAfterCompletion();
-
+		// Should we throw an exception?
+		NamedResultSetMappingMemento memento = resultSetMapping == null ? null : getResultSetMappingMemento( resultSetMapping.getName() );
 		try {
 			// Same approach as AbstractSharedSessionContract#createNativeQuery(String, String)
-			final ReactiveNativeQueryImpl<R> nativeQuery = resultSetMapping != null
-					? new ReactiveNativeQueryImpl<>( queryString, getResultSetMappingMemento( resultSetMapping.getName() ), this )
-					: new ReactiveNativeQueryImpl<>( queryString, this );
+			final ReactiveNativeQueryImpl<R> nativeQuery = new ReactiveNativeQueryImpl<>( queryString, memento, null, this );
 			applyQuerySettingsAndHints( nativeQuery );
 			return nativeQuery;
 		}
@@ -652,20 +724,13 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 		if ( mapping == null ) {
 			throw new IllegalArgumentException( "result set mapping does not exist: " + mappingName );
 		}
-//
-//		ResultSetMappingImpl resultSetMapping = new ResultSetMappingImpl( "impl" );
-//		if ( resultType != null ) {
-//			Class<?> mappedResultType = resultSetMapping.;
-//			if ( !resultType.equals( mappedResultType ) ) {
-//				throw new IllegalArgumentException( "incorrect result type for result set mapping: " + mappingName + " has type " + mappedResultType.getName() );
-//			}
-//		}
 
 		return new ResultSetMapping<>() {
 			@Override
 			public String getName() {
 				return mappingName;
 			}
+
 			@Override
 			public Class<T> getResultType() {
 				return resultType;

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveStatelessSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveStatelessSessionImpl.java
@@ -22,6 +22,7 @@ import org.hibernate.bytecode.spi.BytecodeEnhancementMetadata;
 import org.hibernate.cache.spi.access.EntityDataAccess;
 import org.hibernate.collection.spi.PersistentCollection;
 import org.hibernate.dialect.Dialect;
+import org.hibernate.engine.internal.ReactivePersistenceContextAdapter;
 import org.hibernate.engine.spi.EntityKey;
 import org.hibernate.engine.spi.LoadQueryInfluencers;
 import org.hibernate.engine.spi.PersistenceContext;
@@ -47,10 +48,12 @@ import org.hibernate.persister.entity.EntityPersister;
 import org.hibernate.proxy.HibernateProxy;
 import org.hibernate.proxy.LazyInitializer;
 import org.hibernate.query.IllegalMutationQueryException;
+import org.hibernate.query.UnknownNamedQueryException;
 import org.hibernate.query.criteria.JpaCriteriaInsert;
 import org.hibernate.query.criteria.JpaCriteriaQuery;
 import org.hibernate.query.criteria.JpaRoot;
 import org.hibernate.query.hql.spi.SqmQueryImplementor;
+import org.hibernate.query.named.NamedResultSetMappingMemento;
 import org.hibernate.query.spi.HqlInterpretation;
 import org.hibernate.query.spi.QueryImplementor;
 import org.hibernate.query.sql.spi.NativeQueryImplementor;
@@ -64,7 +67,6 @@ import org.hibernate.query.sqm.tree.select.SqmSelectStatement;
 import org.hibernate.query.sqm.tree.update.SqmUpdateStatement;
 import org.hibernate.reactive.common.AffectedEntities;
 import org.hibernate.reactive.common.ResultSetMapping;
-import org.hibernate.engine.internal.ReactivePersistenceContextAdapter;
 import org.hibernate.reactive.id.ReactiveIdentifierGenerator;
 import org.hibernate.reactive.logging.impl.Log;
 import org.hibernate.reactive.persister.collection.impl.ReactiveCollectionPersister;
@@ -905,7 +907,7 @@ public class ReactiveStatelessSessionImpl extends StatelessSessionImpl implement
 		delayedAfterCompletion();
 
 		try {
-			final ReactiveNativeQueryImpl<R> query = new ReactiveNativeQueryImpl<>( sqlString, this );
+			final ReactiveNativeQueryImpl<R> query = new ReactiveNativeQueryImpl<>( sqlString, null, this);
 			if ( isEmpty( query.getComment() ) ) {
 				query.setComment( "dynamic native SQL query" );
 			}
@@ -956,9 +958,19 @@ public class ReactiveStatelessSessionImpl extends StatelessSessionImpl implement
 		delayedAfterCompletion();
 
 		try {
-			return isNotEmpty( resultSetMappingName )
-					? new ReactiveNativeQueryImpl<>( sqlString, getResultSetMappingMemento( resultSetMappingName ), this )
-					: new ReactiveNativeQueryImpl<>( sqlString, this );
+			if ( isNotEmpty( resultSetMappingName ) ) {
+				final NamedResultSetMappingMemento resultSetMappingMemento = getFactory().getQueryEngine()
+						.getNamedObjectRepository()
+						.getResultSetMappingMemento( resultSetMappingName );
+
+				if ( resultSetMappingMemento == null ) {
+					throw new HibernateException( "Could not resolve specified result-set mapping name : " + resultSetMappingName );
+				}
+				return new ReactiveNativeQueryImpl<>( sqlString, resultSetMappingMemento, null, this );
+			}
+			else {
+				return new ReactiveNativeQueryImpl<>( sqlString, this );
+			}
 			//TODO: why no applyQuerySettingsAndHints( query ); ???
 		}
 		catch (RuntimeException he) {
@@ -1070,8 +1082,61 @@ public class ReactiveStatelessSessionImpl extends StatelessSessionImpl implement
 	}
 
 	@Override
+	public  <R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String queryName) {
+		checksBeforeQueryCreation();
+		try {
+			return (ReactiveQueryImplementor<R>) buildNamedQuery(
+					queryName,
+					this::createSqmQueryImplementor,
+					this::createNativeQueryImplementor
+			);
+		}
+		catch (RuntimeException e) {
+			throw convertNamedQueryException( e );
+		}
+	}
+
+	@Override
+	public <R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String queryName, Class<R> resultType) {
+		checksBeforeQueryCreation();
+		if ( resultType == null ) {
+			throw new IllegalArgumentException( "Result class is null" );
+		}
+		try {
+			return (ReactiveQueryImplementor<R>) buildNamedQuery(
+					queryName,
+					memento -> createSqmQueryImplementor( resultType, memento ),
+					memento -> createNativeQueryImplementor( resultType, memento )
+			);
+		}
+		catch (RuntimeException e) {
+			throw convertNamedQueryException( e );
+		}
+	}
+
+	private RuntimeException convertNamedQueryException(RuntimeException e) {
+		if ( e instanceof UnknownNamedQueryException ) {
+			// JPA expects this to mark the transaction for rollback only
+			getTransactionCoordinator().getTransactionDriverControl().markRollbackOnly();
+			// it also expects an IllegalArgumentException, so wrap UnknownNamedQueryException
+			return new IllegalArgumentException( e.getMessage(), e );
+		}
+		else if ( e instanceof IllegalArgumentException ) {
+			return e;
+		}
+		else {
+			return getExceptionConverter().convert( e );
+		}
+	}
+
+	@Override
 	public <R> ReactiveSelectionQuery<R> createNamedReactiveSelectionQuery(String queryName, Class<R> expectedResultType) {
 		return (ReactiveSelectionQuery<R>) createNamedSelectionQuery( queryName , expectedResultType );
+	}
+
+	private void checksBeforeQueryCreation() {
+		checkOpen();
+		checkTransactionSynchStatus();
 	}
 
 	@Override
@@ -1081,11 +1146,6 @@ public class ReactiveStatelessSessionImpl extends StatelessSessionImpl implement
 			throw new IllegalMutationQueryException( "Expecting a native mutation query, but found `" + sqlString + "`" );
 		}
 		return query;
-	}
-
-	@Override
-	public <R> ReactiveQueryImplementor<R> createReactiveNamedQuery(String queryName, Class<R> resultType) {
-		return (ReactiveQueryImplementor<R>) buildNamedQuery( queryName, resultType );
 	}
 
 	@Override
@@ -1131,11 +1191,13 @@ public class ReactiveStatelessSessionImpl extends StatelessSessionImpl implement
 		pulseTransactionCoordinator();
 		delayedAfterCompletion();
 
+		if ( resultSetMapping == null ) {
+			throw new IllegalArgumentException( "Result set mapping was not specified" );
+		}
+
 		try {
-			// Same approach as AbstractSharedSessionContract#createNativeQuery(String, String)
-			final ReactiveNativeQueryImpl<R> nativeQuery = resultSetMapping != null
-					? new ReactiveNativeQueryImpl<>( queryString, getResultSetMappingMemento( resultSetMapping.getName() ), this )
-					: new ReactiveNativeQueryImpl<>( queryString, this );
+			final NamedResultSetMappingMemento memento = getResultSetMappingMemento( resultSetMapping.getName() );
+			final ReactiveNativeQueryImpl<R> nativeQuery = new ReactiveNativeQueryImpl<>( queryString, memento, null, this );
 			applyQuerySettingsAndHints( nativeQuery );
 			return nativeQuery;
 		}

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/impl/StageSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/impl/StageSessionImpl.java
@@ -5,16 +5,11 @@
  */
 package org.hibernate.reactive.stage.impl;
 
-import jakarta.persistence.CacheRetrieveMode;
-import jakarta.persistence.CacheStoreMode;
-import jakarta.persistence.EntityGraph;
-import jakarta.persistence.FlushModeType;
-import jakarta.persistence.LockModeType;
-import jakarta.persistence.PersistenceException;
-import jakarta.persistence.criteria.CriteriaDelete;
-import jakarta.persistence.criteria.CriteriaQuery;
-import jakarta.persistence.criteria.CriteriaUpdate;
-import jakarta.persistence.metamodel.Attribute;
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
 import org.hibernate.CacheMode;
 import org.hibernate.Filter;
 import org.hibernate.FlushMode;
@@ -39,10 +34,16 @@ import org.hibernate.reactive.stage.Stage.MutationQuery;
 import org.hibernate.reactive.stage.Stage.Query;
 import org.hibernate.reactive.stage.Stage.SelectionQuery;
 
-import java.lang.invoke.MethodHandles;
-import java.util.List;
-import java.util.concurrent.CompletionStage;
-import java.util.function.Function;
+import jakarta.persistence.CacheRetrieveMode;
+import jakarta.persistence.CacheStoreMode;
+import jakarta.persistence.EntityGraph;
+import jakarta.persistence.FlushModeType;
+import jakarta.persistence.LockModeType;
+import jakarta.persistence.PersistenceException;
+import jakarta.persistence.criteria.CriteriaDelete;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.CriteriaUpdate;
+import jakarta.persistence.metamodel.Attribute;
 
 import static org.hibernate.reactive.util.impl.CompletionStages.applyToAll;
 import static org.hibernate.reactive.util.impl.CompletionStages.returnOrRethrow;
@@ -556,7 +557,7 @@ public class StageSessionImpl implements Stage.Session {
 
 	@Override
 	public <R> Query<R> createNamedQuery(String queryName) {
-		return new StageQueryImpl<>( delegate.createReactiveNamedQuery( queryName, null ) );
+		return new StageQueryImpl<>( delegate.createReactiveNamedQuery( queryName ) );
 	}
 
 	@Override

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/impl/StageStatelessSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/impl/StageStatelessSessionImpl.java
@@ -326,7 +326,7 @@ public class StageStatelessSessionImpl implements Stage.StatelessSession {
 
 	@Override
 	public <R> Query<R> createNamedQuery(String queryName) {
-		return new StageQueryImpl<>( delegate.createReactiveNamedQuery( queryName, null ) );
+		return new StageQueryImpl<>( delegate.createReactiveNamedQuery( queryName ) );
 	}
 
 	@Override

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveStatelessWithBatchTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveStatelessWithBatchTest.java
@@ -3,11 +3,10 @@
  * SPDX-License-Identifier: Apache-2.0
  * Copyright: Red Hat Inc. and Hibernate Authors
  */
-package org.hibernate;
+package org.hibernate.reactive;
 
 import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
 import org.hibernate.cfg.Configuration;
-import org.hibernate.reactive.BaseReactiveTest;
 import org.hibernate.reactive.annotations.EnabledFor;
 import org.hibernate.reactive.testing.SqlStatementTracker;
 

